### PR TITLE
STOR-2002: add Azure File snapshot support

### PR DESCRIPTION
**Pull request by RomanBednar**
## Basic verification for snapshot create and restore

Driver logs with successful PVC and snapshot creation:
```
I0802 11:27:44.523693       1 utils.go:101] GRPC call: /csi.v1.Controller/CreateVolume
I0802 11:27:44.523715       1 utils.go:102] GRPC request: {"capacity_range":{"required_bytes":1073741824},"name":"pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6","parameters":{"csi.storage.k8s.io/pv/name":"pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6","csi.storage.k8s.io/pvc/name":"pvc-1","csi.storage.k8s.io/pvc/namespace":"default","skuName":"Standard_LRS"},"volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4","mount_flags":["mfsymlinks","cache=strict","nosharesock","actimeo=30"]}},"access_mode":{"mode":7}}]}
I0802 11:27:45.199848       1 controllerserver.go:561] begin to create file share(pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6) on account(clusterewxps) type(Standard_LRS) subID() rg(rbednar-mycluster-01-wnmgz-rg) location() size(1) protocol(SMB)
I0802 11:27:45.586538       1 controllerserver.go:593] create file share pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6 on storage account clusterewxps successfully
I0802 11:27:45.607282       1 controllerserver.go:639] store account key to k8s secret(azure-storage-account-clusterewxps-secret) in default namespace
I0802 11:27:45.607340       1 utils.go:108] GRPC response: {"volume":{"capacity_bytes":1073741824,"volume_context":{"csi.storage.k8s.io/pv/name":"pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6","csi.storage.k8s.io/pvc/name":"pvc-1","csi.storage.k8s.io/pvc/namespace":"default","secretnamespace":"default","skuName":"Standard_LRS"},"volume_id":"rbednar-mycluster-01-wnmgz-rg#clusterewxps#pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6###default"}}
I0802 11:28:00.070584       1 utils.go:101] GRPC call: /csi.v1.Identity/GetPluginInfo
I0802 11:28:00.070604       1 utils.go:102] GRPC request: {}
I0802 11:28:00.070645       1 utils.go:108] GRPC response: {"name":"file.csi.azure.com","vendor_version":"v1.30.3"}
I0802 11:28:00.071131       1 utils.go:101] GRPC call: /csi.v1.Controller/CreateSnapshot
I0802 11:28:00.071144       1 utils.go:102] GRPC request: {"name":"snapshot-9cde69ce-2e33-4f32-b4ba-471b04549904","source_volume_id":"rbednar-mycluster-01-wnmgz-rg#clusterewxps#pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6###default"}
I0802 11:28:00.414963       1 controllerserver.go:915] created share snapshot: 2024-08-02T11:28:00.0000000Z, time: 2024-08-02 11:28:00 +0000 UTC, quota: 0GiB
I0802 11:28:00.415040       1 utils.go:108] GRPC response: {"snapshot":{"creation_time":{"seconds":1722598080},"ready_to_use":true,"size_bytes":107374182400,"snapshot_id":"rbednar-mycluster-01-wnmgz-rg#clusterewxps#pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6###default#2024-08-02T11:28:00.0000000Z","source_volume_id":"rbednar-mycluster-01-wnmgz-rg#clusterewxps#pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6###default"}}
```

Driver logs during snapshot restore:
```
I0802 11:29:33.283200       1 controllerserver.go:1098] copied fileshare pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6(snapshot: 2024-08-02T11:28:00.0000000Z) to pvc-10e9e706-2427-46c4-b988-bf7aa228e718 successfully
I0802 11:29:33.283262       1 controllerserver.go:593] create file share pvc-10e9e706-2427-46c4-b988-bf7aa228e718 on storage account clusterewxps successfully
I0802 11:29:33.283302       1 utils.go:108] GRPC response: {"volume":{"capacity_bytes":107374182400,"content_source":{"Type":{"Snapshot":{"snapshot_id":"rbednar-mycluster-01-wnmgz-rg#clusterewxps#pvc-ce4d6eca-df1c-41dd-bb13-a484c0524ec6###default#2024-08-02T11:28:00.0000000Z"}}},"volume_context":{"csi.storage.k8s.io/pv/name":"pvc-10e9e706-2427-46c4-b988-bf7aa228e718","csi.storage.k8s.io/pvc/name":"pvc-1-snapshot-restore","csi.storage.k8s.io/pvc/namespace":"default","secretnamespace":"default","skuName":"Standard_LRS"},"volume_id":"rbednar-mycluster-01-wnmgz-rg#clusterewxps#pvc-10e9e706-2427-46c4-b988-bf7aa228e718###default"}}
```

Snapshot restored successfully:
```
 oc get pvc/pvc-1-snapshot-restore
NAME                     STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS    VOLUMEATTRIBUTESCLASS   AGE
pvc-1-snapshot-restore   Bound    pvc-10e9e706-2427-46c4-b988-bf7aa228e718   100Gi      RWO            azurefile-csi   <unset>                 2m23s
```

cc @openshift/storage

### CODE REVIEW

File: assets/overlays/azure-file/generated/hypershift/controller.yaml [hunk 1]
```diff
@@ -169,8 +173,14 @@ spec:
         - --leader-election-retry-period=${LEADER_ELECTION_RETRY_PERIOD}
         - --leader-election-namespace=openshift-cluster-csi-drivers
         - --v=${LOG_LEVEL}
+        - --default-fstype=ext4
+        - --feature-gates=Topology=true
         - --extra-create-metadata=true
-        - --timeout=300s
+        - --strict-topology=true
+        - --timeout=30s
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply assets/overlays/azure-file/generated/hypershift/controller.yaml]: 1200s seems a lot, I've set 300s now and we can increase it later if we see issues.

File: assets/overlays/azure-file/generated/hypershift/controller.yaml [hunk 2]
```diff
@@ -169,8 +173,14 @@ spec:
         - --leader-election-retry-period=${LEADER_ELECTION_RETRY_PERIOD}
         - --leader-election-namespace=openshift-cluster-csi-drivers
         - --v=${LOG_LEVEL}
+        - --default-fstype=ext4
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply assets/overlays/azure-file/generated/hypershift/controller.yaml]: Right, it should be left unset - removing.

File: assets/overlays/azure-file/generated/hypershift/controller.yaml [hunk 3]
```diff
@@ -169,8 +173,14 @@ spec:
...
+        - --feature-gates=Topology=true
         - --extra-create-metadata=true
-        - --timeout=300s
+        - --strict-topology=true
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply assets/overlays/azure-file/generated/hypershift/controller.yaml]: That's a mistake, removing.

**Comment by RomanBednar** [comment]
The problem with snapshot restore size turned out to be a bug in the driver. Upstream posted a fix real quick and I've re-tested the failing CI tests with that fix and they passed for me: https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/2040

So we need to wait for upstream merge, follow with a cherry-pick and we should be good with this PR.

**Comment by RomanBednar** [comment]
cc @jsafrane for review

File: pkg/driver/azure-file/azure_file.go [hunk 4]
```diff
@@ -65,9 +65,12 @@ func GetAzureFileGeneratorConfig() *generator.CSIDriverGeneratorConfig {
 			SidecarLocalMetricsPortStart:   commongenerator.AzureFileLoopbackMetricsPortStart + 1,
 			SidecarExposedMetricsPortStart: commongenerator.AzureFileExposedMetricsPortStart + 1,
 			Sidecars: []generator.SidecarConfig{
-				commongenerator.DefaultProvisioner.WithExtraArguments(
+				commongenerator.DefaultProvisionerWithSnapshots.WithExtraArguments(
 					"--extra-create-metadata=true",
 					"--timeout=300s",
+					"--worker-threads=100",
```
Comment (jsafrane) [inline pkg/driver/azure-file/azure_file.go:71]: The upstream Deployment does not set `--worker-threads`

Comment (RomanBednar) [reply pkg/driver/azure-file/azure_file.go]: Removed.

**Comment by jsafrane** [comment]
/retest-required
Azure tests have issues pulling from registry.build09.ci.openshift.org !?

**Comment by duanwei33** [comment]
Many CI failed due to two cases, so just retest.
Run multi-stage test pre phase 
Run multi-stage test post phase 

/retest-required

**Pull request merged**